CORE-8688: Tracking OOM's in crash tracker

michael-redpanda opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

This adds support for tracking OOMs in the crash tracker.  The OOM callback function in `memory_sampling.cc` was modified to also start writing messages to the crash tracker when it is called by Seastar during an OOM.

* The crash recorder was modified to provide `reset()` functionality to be used in tests
* The recorder now has methods that returns a function to be called to append messages to the crash log for OOM's
* The OOM callback in `memory_sampling.cc` will now call into the crash recorder and obtain that function and write its messages out to the crash log
* Tests were added

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

* None

michael-redpanda committed:
crash-tracker: Add OOM writing functions

Added methods to be used when OOM crashes are occuring.  It provides a
function that can be used to append multiple messages to the crash
message.  Added a finish method that performs the write.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
application: Moved up creation of crash tracker

Moved crash tracker initialization to before initialize().  The crash
tracker service will be needed by the memory_sampling serivce to record
OOM messages.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
resource_mgmt: Logging OOM messages to crash log

When the get_oom_diagnostics_callback() method is called when Seastar is
exiting due to an OOM, it will attempt to log messages to the crash
logger.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
crash_tracker: Provide reset semantics

For tests that require using the static crash recorder, add methods that
permit resetting the crash tracker so it can be reused between tests.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
crash_tracker: Tidy up reserved_string

Add some types and fix some return types

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
crash_tracker/t: Added test for OOM messaging

Signed-off-by: Michael Boquard <michael@redpanda.com>

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--716efb0e-12cb-4d58-a160-4d5764e1a6e7-->
<!--TEST RESULTS-->
<!--e29d8b28-15ff-4210-99fb-7b7b61709cb3-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62472">build#62472</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.compaction_test.CompactionTest.test_compaction.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/62472#01954e8f-659c-4582-ba61-5e31d8cd390b | FLAKY | 1/3 |
| rptest.tests.datalake.compaction_test.CompactionTest.test_compaction.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/62472#01954e8f-659d-4354-8ca8-359fea71bfec | FLAKY | 1/2 |
| rptest.tests.full_disk_test.LogStorageMaxSizeSI.test_stay_below_target_size.log_segment_size=1048576.cleanup_policy=compact | ducktape | https://buildkite.com/redpanda/redpanda/builds/62472#01954e8c-2e9e-4e67-8629-e24948e53f8f | FLAKY | 1/2 |
| rptest.tests.log_segment_ms_test.SegmentMsTest.test_segment_rolling.use_alter_cfg=True.server_cfg=None.topic_cfg=300000 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62472#01954e8c-2e9d-437d-97b8-62bda86439b0 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62594">build#62594</a></summary>


| test_id                                                                                                                                                                                                                                            | test_kind   | job_url                                                                                   | test_status   | passed   |
|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_restore_partition_spec.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP                                                                      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62594#019561a9-1c68-43b1-ad26-5bcc1a557837 | FLAKY         | 1/2      |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_restore_partition_spec.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC                                                                        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62594#019561bc-4eb8-404e-987c-19a4f37f6f5e | FLAKY         | 1/2      |
| rptest.tests.datalake.datalake_e2e_test.DatalakeE2ETests.test_topic_lifecycle.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE                                                                                               | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62594#019561bc-4eb7-4680-8552-46de9d3d37b9 | FLAKY         | 1/2      |
| rptest.tests.full_disk_test.LogStorageMaxSizeSI.test_stay_below_target_size.log_segment_size=1048576.cleanup_policy=compact                                                                                                                        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62594#019561a9-1c68-43b1-ad26-5bcc1a557837 | FLAKY         | 1/2      |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=False.with_tiered_storage=True.with_iceberg=False.with_chunked_compaction=False.cloud_storage_type=CloudStorageType.ABS | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62594#019561a9-1c69-4e01-87bc-f5f6b7e3e795 | FLAKY         | 1/2      |
| rptest.tests.upgrade_test.UpgradeWithWorkloadTest.test_rolling_upgrade_with_rollback.upgrade_after_rollback=True                                                                                                                                   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62594#019561bc-4eb8-404e-987c-19a4f37f6f5e | FLAKY         | 1/2      |
| storage_e2e_single_thread_rpunit.storage_e2e_single_thread_rpunit                                                                                                                                                                                  | unit        | https://buildkite.com/redpanda/redpanda/builds/62594#01956161-0562-4f16-9900-ff274413630a | FLAKY         | 1/2      |
</details>
<!--/body-->

dotnwat commented on src/v/crash_tracker/types.h:
```diff
@@ -40,7 +40,12 @@ struct reserved_string {
     static_assert(Size > 0, "Size must be greater than 0");
```
> struct reserved_string

just curious about this reserved_string type... does it do something that wouldn't be possible by using a `seastar::basic_sstring` with a larger SSO reservation value?

michael-redpanda replied:
We opted for something that would prevent a developer from accidentally introducing code that would cause an allocation, say if they attempt to create a string >4096 bytes.

StephanDollberg commented on src/v/resource_mgmt/memory_sampling.cc:
```diff
@@ -64,6 +67,51 @@ memory_sampling::get_oom_diagnostics_callback() {
              seastar::memory::memory_diagnostics_writer writer) mutable {
+        auto& recorder = crash_tracker::get_recorder();
+        auto oom_recorder = recorder.begin_oom_recording();
+        auto finish_oom_writing = ss::defer([&recorder, &oom_recorder] {
```
Does `ss::defer` not allocate? All of this needs to be allocation free.

dotnwat replied:
>Does ss::defer not allocate? All of this needs to be allocation free.

oh i didn't realize this was invoked inline with malloc like the batch cache via the reclaimer. i guess that means this code is even a little more sensitive than just avoiding malloc and needs to be thought about as if its execution preempted the reactor thread, so the normal scheduling point analysis of concurrency doesn't quite apply?

michael-redpanda replied:
> Does ss::defer not allocate? All of this needs to be allocation free.

I don't believe so.  The [`test_no_allocs_in_oom_callback`](https://github.com/redpanda-data/redpanda/blob/de6bc7e65e9fd14bb07ba8406d6929b6dbfce043/src/v/resource_mgmt/tests/memory_sampling_tests.cc#L31-L74) test passes

pgellert replied:
It looks like `ss::defer()` just moves / internalises the lambda without heap allocation: https://github.com/scylladb/seastar/blob/3ee9204cb792885c4a0bdb19a61619572d88c963/include/seastar/util/defer.hh#L46-L76

dotnwat replied:
the no-alloc property is important, but don't forget the preemptive nature too. it needs to be safe w.r.t. any code that may have been running, called malloc internally, and then ended up in the callback.

pgellert commented on src/v/crash_tracker/recorder.cc:
```diff
@@ -280,6 +280,32 @@ void recorder::record_crash_vassert(std::string_view msg) {
+    return [this](std::string_view ms) {
+        vassert(_oom_writer.has_value(), "OOM message already recorded");
+        auto& oom_writer = _oom_writer.value();
+        oom_writer._oom_msg_pos = fmt::format_to(
+          oom_writer._oom_msg_pos, "{}", ms);
```
I think if we write more than the crash message's reserved size then this could be unsafe. I think we need to make sure we only write up to the `crash_message.begin()` iterator here.

nit: I'm also wondering if it would be simpler to turn this lambda into a method on `oom_writer`?

michael-redpanda replied:
Good points all around

michael-redpanda replied:
> make sure we only write up to the crash_message.begin()

I believe you meant `crash_message.end()` right?

pgellert commented on src/v/crash_tracker/prepared_writer.h:
```diff
@@ -51,6 +51,9 @@ class prepared_writer {
     /// Must be called after a fill() that returned a non-null value
     void write();
 
+    /// Only to be used under testing situations
```
nit: it might be worth calling out in the doc comment on `prepared_writer` (at the top of this file) that `reset()` exists for testing and that it moves any state back to `uninitialized`.

pgellert commented on src/v/crash_tracker/prepared_writer.cc:
```diff
@@ -187,4 +187,10 @@ ss::future<> prepared_writer::release() {
+    _state = state::uninitialized;
+    _serde_output.clear();
+    _prepared_cd = crash_description{};
+}
```
Do we need to reset `_crash_report_file_name` and `_fd` as well?

michael-redpanda replied:
We could, though with this method you will have to call `initialize()` which resets those two anyways

michael-redpanda replied:
~~I'll throw an assert in to ensure that it's in the released state.~~

That won't work.  I'll clean up things

pgellert commented on src/v/resource_mgmt/memory_sampling.cc:
```diff
@@ -12,15 +12,18 @@
 #include "resource_mgmt/available_memory.h"
 #include "ssx/future-util.h"
 #include "ssx/sformat.h"
+#include "utils/human.h"
```
Love this header name. Makes you think that you're importing a human here.

pgellert commented on src/v/resource_mgmt/memory_sampling.cc:
```diff
@@ -64,6 +67,51 @@ memory_sampling::get_oom_diagnostics_callback() {
+            auto free_mem = stats.free_memory();
+            auto failed_allocs = stats.failed_allocations();
+            (*oom_recorder)("Dumping seastar memory diagnostics\n");
+            auto bytes_written = fmt::format_to_n(
+                                   format_buf.begin(),
+                                   format_buf.size(),
+                                   "Used memory:   {}\n",
+                                   human::bytes(
+                                     static_cast<double>(total_mem - free_mem)))
+                                   .size;
+            (*oom_recorder)(std::string_view(format_buf.data(), bytes_written));
```
`fmt::format_to_n` has a bit of a gotcha here where `format_to_n_result.size` contains the *bytes that you wanted to print* not the *bytes that got printed*. So I think we either need to use `format_to_n_result.out` here or `std::min()` with `format_buf.size()`.

```
diff --git i/src/v/crash_tracker/tests/types_test.cc w/src/v/crash_tracker/tests/types_test.cc
index fd9457cd2f..a1df2e8f38 100644
--- i/src/v/crash_tracker/tests/types_test.cc
+++ w/src/v/crash_tracker/tests/types_test.cc
@@ -49,7 +49,11 @@ TEST_F(TestReservedStringSerde, BelowCapacity) {
 }

 TEST_F(TestReservedStringSerde, AboveCapacity) {
-    run_test_case("123456", "12345");
+    auto rs_in = ss::sstring(3, '\0');
+    auto res = fmt::format_to_n(rs_in.data(), rs_in.size(), "{}", "12345");
+    ASSERT_EQ(res.size, 5);
+    ASSERT_EQ(std::string{rs_in}, std::string{"123"});
+    ASSERT_EQ(res.out, rs_in.end());
 }

 } // namespace crash_tracker

```

michael-redpanda replied:
good catch, thanks

pgellert commented on src/v/resource_mgmt/memory_sampling.cc:
```diff
@@ -64,6 +67,51 @@ memory_sampling::get_oom_diagnostics_callback() {
     return [allocation_sites = std::move(allocation_sites),
             format_buf = std::move(format_buf)](
              seastar::memory::memory_diagnostics_writer writer) mutable {
+        auto& recorder = crash_tracker::get_recorder();
+        auto oom_recorder = recorder.begin_oom_recording();
```
There is an implicit dependency here that the crash recorder must be initialized before the oom callback is registered. I think this might be the best place to add a comment to call this dependency out (putting it into `application.cc` seems out of place).

michael-redpanda commented:
Force push [cdf53d2](https://github.com/redpanda-data/redpanda/commit/cdf53d252cabd85d31d8d92c76b784b5f30accc0):

* Addressed PR comments

michael-redpanda commented:
Force push [215b1e2](https://github.com/redpanda-data/redpanda/commit/215b1e2be859092709880cca70355ce419e9cf67):

* Rebased off of dev and added stringification of `crash_type::oom` to `src/v/crash_tracker/types.cc`

pgellert reviewed (approved):
lgtm

dotnwat commented on src/v/resource_mgmt/memory_sampling.cc:
```diff
@@ -64,6 +67,57 @@ memory_sampling::get_oom_diagnostics_callback() {
+                                   format_buf.begin(),
+                                   format_buf.size(),
+                                   "Used memory:   {}\n",
+                                   human::bytes(
+                                     static_cast<double>(total_mem - free_mem)))
```
would human::bytes allocate and would that be a problem? it looks like it probably would given it's formatting into std::ostream?

michael-redpanda replied:
yeah good catch... I wonder if fmtlib is providing the iterator as the stream to `operator<<`.  I'd think removing the `operator<<` and replacing it with a `fmt::formatter` would be more appropriate.  This is what happens with`ss::memory::allocation_site`: https://github.com/redpanda-data/redpanda/blob/1128287e89d5c93362aa858e868a48a7a8a3259c/src/v/resource_mgmt/memory_sampling.cc#L37-L46

michael-redpanda replied:
We are seeing this pattern in Seastar as well: https://github.com/scylladb/seastar/blob/9cb6a258c47f464b89e47b1a962945d943f47910/src/core/memory.cc#L1989-L2035

This is also called during an OOM event

michael-redpanda replied:
I made a small reproducer, it looks like it takes the OutputIterator passed to `format_to_n` and converts it to a stream.  This is call right before the `ostream<<`:

```c++
template <typename Char, typename T>
void format_value(buffer<Char>& buf, const T& value,
                  locale_ref loc = locale_ref()) {
  auto&& format_buf = formatbuf<std::basic_streambuf<Char>>(buf);
  auto&& output = std::basic_ostream<Char>(&format_buf);
#if !defined(FMT_STATIC_THOUSANDS_SEPARATOR)
  if (loc) output.imbue(loc.get<std::locale>());
#endif
  output << value;
  output.exceptions(std::ios_base::failbit | std::ios_base::badbit);
}
```

Looks like it turns the buffer into a stream_buf then into an ostream.

michael-redpanda replied:
I also want to point out that the [`test_no_allocs_in_oom_callback`](https://github.com/redpanda-data/redpanda/blob/56a662110414fdb4ceeb9ed6238ddf233aeb908f/src/v/resource_mgmt/tests/memory_sampling_tests.cc#L31-L74) test passes.  That does look at the number of allocs before and after the OOM callback and if they don't match the test fails.
